File: backend/app/schemas/Classwork.py
# app/schemas/Classwork.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime


class ClassworkCreate(BaseModel):
    title: str
    description: Optional[str] = None
    instructions: Optional[str] = None
    classwork_type: str  # QUIZ, ASSIGNMENT, ACTIVITY
    classwork_category: Optional[str] = None  # WRITTEN_WORK, PERFORMANCE_TASK, PERIODICAL_EXAM
    total_points: Optional[float] = 100
    subject_id: int
    is_published: Optional[bool] = False
    lesson_ids: Optional[list[int]] = None


class ClassworkUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    instructions: Optional[str] = None
    classwork_type: Optional[str] = None
    classwork_category: Optional[str] = None
    total_points: Optional[float] = None
    is_published: Optional[bool] = None
    lesson_ids: Optional[list[int]] = None


class ClassworkAttachmentResponse(BaseModel):
    classwork_attachment_id: int
    file_name: str
    file_type: Optional[str]
    file_size: int
    uploaded_at: Optional[datetime]


class CwAssignmentRow(BaseModel):
    classwork_assignment_id: int
    classwork_id: int
    class_id: int
    title: Optional[str] = None
    classwork_type: Optional[str] = None
    due_date: Optional[datetime] = None
    is_published: bool


class ClassworkResponse(BaseModel):
    classwork_id: int
    title: str
    description: Optional[str]
    instructions: Optional[str]
    classwork_type: str
    classwork_category: Optional[str]
    total_points: Optional[float]
    is_published: bool
    is_locked: bool
    subject_id: int
    subject_name: Optional[str] = None
    created_by_staff_id: str
    teacher_name: Optional[str] = None
    attachments: list[ClassworkAttachmentResponse] = []
    assignments: Optional[list[CwAssignmentRow]] = None
    created_at: Optional[datetime]
    updated_at: Optional[datetime]


class ClassworkAssignRequest(BaseModel):
    class_ids: list[int]
    publish_date: Optional[datetime] = None
    due_date: Optional[datetime] = None
    lock_date: Optional[datetime] = None
    max_attempts: Optional[int] = 1
    is_published: Optional[bool] = True


class ClassworkAssignmentResponse(BaseModel):
    """Class-specific classwork view used by student and teacher assignment pages."""
    classwork_assignment_id: int
    classwork_id: int
    class_id: int
    section_name: Optional[str] = None
    title: str
    description: Optional[str]
    instructions: Optional[str]
    classwork_type: str
    classwork_category: Optional[str]
    total_points: Optional[float]
    due_date: Optional[datetime]
    is_published: bool
    is_locked: Optional[bool] = False
    max_attempts: Optional[int] = 1
    teacher_name: Optional[str] = None
    attachments: list[ClassworkAttachmentResponse] = []
    assignments: Optional[list[CwAssignmentRow]] = None
    submission_status: Optional[str] = None  # for student view

File: frontend/src/types/classwork.ts
export type ClassworkType = "READING" | "ACTIVITY" | "ASSIGNMENT" | "QUIZ" | string;
export type ClassworkKind = "READING" | "ACTIVITY" | "ASSIGNMENT" | "QUIZ";
export type TabId = "all" | "readings" | "activities" | "assignments" | "quizzes";
export type SortMode = "newest" | "oldest" | "title";
export type CreateStep = "type" | "quiz-source" | "details" | "quiz" | "assign";

export type ClassworkAttachment = {
  classwork_attachment_id: number;
  file_name: string;
  file_type?: string;
  file_size: number;
  uploaded_at?: string | null;
};

export type ClassworkAssignment = {
  classwork_assignment_id: number;
  classwork_id: number;
  class_id: number;
  title?: string | null;
  due_date?: string | null;
  lock_date?: string | null;
  allow_late_submissions?: boolean;
  max_attempts?: number | null;
  is_published: boolean;
  is_locked?: boolean | null;
};

export type TeacherClasswork = {
  classwork_id: number;
  title: string;
  description?: string | null;
  instructions?: string | null;
  classwork_type: ClassworkType;
  classwork_category?: string | null;
  total_points?: number | null;
  is_published: boolean;
  is_locked: boolean;
  is_archived: boolean;
  subject_id: number;
  subject_name?: string | null;
  attachments: ClassworkAttachment[];
  assignments?: ClassworkAssignment[] | null;
  created_at?: string | null;
  updated_at?: string | null;
};

export type TeacherClassLoad = {
  subject_load_id: number;
  subject_id: number;
  subject_name: string;
  subject_codename?: string | null;
  class_id: number;
  section_name: string;
};

export type TeacherLesson = {
  lesson_id: number;
  title: string;
  order_index?: number;
  is_published: boolean;
  is_draft: boolean;
  subject_id: number;
};

export type TrackingStudent = {
  student_id: string;
  student_name: string;
  status: string;
  submission_id?: number | null;
  grade?: number | null;
  attachment_count?: number;
};

export type AssignmentTracking = {
  classwork_assignment_id: number;
  classwork_id: number;
  classwork_title?: string | null;
  total_students: number;
  submitted_count: number;
  missing_count: number;
  submitted: TrackingStudent[];
  missing: TrackingStudent[];
};

export type SubmissionAttachment = {
  submission_attachment_id: number;
  file_name: string;
  file_type?: string | null;
  file_size: number;
  uploaded_at?: string | null;
};

export type TeacherSubmissionDetail = {
  submission_id: number;
  student_id: string;
  student_name?: string | null;
  classwork_assignment_id: number;
  status: string;
  grade?: number | null;
  feedback?: string | null;
  submitted_at?: string | null;
  attachments: SubmissionAttachment[];
  total_points?: number | null;
};

export type CreateDraft = {
  subject_id: string;
  title: string;
  description: string;
  instructions: string;
  classwork_category: string;
  total_points: string;
  due_date: string;
  lock_date: string;
  allow_late_submissions: boolean;
  max_attempts: string;
  is_published: boolean;
};

export type EditDraft = {
  title: string;
  description: string;
  instructions: string;
  classwork_type: string;
  classwork_category: string;
  total_points: string;
  due_date: string;
  lock_date: string;
  allow_late_submissions: boolean;
  max_attempts: string;
  is_published: boolean;
};
